packages/router/src/history/liveview.rs

```diff
@@ -0,0 +1,443 @@
+                            let mut timeline = timeline.lock().expect("unpoisoned mutex");
+                            let state = timeline.push(route.clone());
+                            let state = serde_json::to_string(&state).expect("serializable state");
+                            let session = serde_json::to_string(&timeline.session())
+                                .expect("serializable session");
+                            create_eval(&format!(
+                                r#"
+                                // this does not trigger a PopState event
+                                history.pushState({state}, "", "{route}");
+                                sessionStorage.setItem("liveview", '{session}');
+                            "#
+                            ))
+                        }
+                        Action::Replace(route) => {
+                            let mut timeline = timeline.lock().expect("unpoisoned mutex");
+                            let state = timeline.replace(route.clone());
+                            let state = serde_json::to_string(&state).expect("serializable state");
+                            let session = serde_json::to_string(&timeline.session())
+                                .expect("serializable session");
+                            create_eval(&format!(
+                                r#"
+                                // this does not trigger a PopState event
+                                history.replaceState({state}, "", "{route}");
+                                sessionStorage.setItem("liveview", '{session}');
+                            "#
+                            ))
+                        }
+                        Action::External(url) => create_eval(&format!(
+                            r#"
+                                location.href = "{url}";
+                            "#
+                        )),
+                    };
+                }
+            }
+        });
+
+        // Listen to browser actions
+        let window = cx.consume_context::<Window>().unwrap();
+        cx.push_future({
+            let mut window_rx = window.subscribe();
+            let updater = updater_callback.clone();
+            let timeline = timeline.clone();
+            let create_eval = create_eval.clone();
+            async move {
+                loop {
+                    let window_event = window_rx.recv().await.expect("sender to exist");
+                    match window_event {
+                        WindowEvent::Load {
+                            location,
+                            state,
+                            session,
+                            depth,
+                        } => {
+                            let Ok(route) = R::from_str(&location.to_string()) else {
+                                return;
+                            };
+                            let Ok(state) = serde_json::from_str(&state) else {
```

Comment:
Instaed of storing the current history in each history in the browser history, could we just serialize any part of the history with javascript that looks like it could be from the same applicateion (any urls that have the same domain?).

Reply:
This could prove to be hard, as security restrictions in browsers hinder programmatically getting any other information other than the state of the current page and the total number of pages in the stack. However, I think for this particular problem, with efficiency in mind, we can probably scale down the amount of information being sent drastically.

Reply:
This could prove to be hard, as security restrictions in browsers hinder programmatically getting any other information other than the state of the current page and the total number of pages in the stack. However, I think for this particular problem, with efficiency in mind, we can probably scale down the amount of information being sent drastically. I will investigate this further.

Reply:
> This could prove to be hard, as security restrictions in browsers hinder programmatically getting any other information other than the state of the current page and the total number of pages in the stack. However, I think for this particular problem, with efficiency in mind, we can probably scale down the amount of information being sent drastically. I will investigate this further.

That makes sense, since it only transfers the history on the initial load, this shouldn't be much of a performance hit